tests/test_main.cc

```diff
@@ -0,0 +1,154 @@
+#include <gtest/gtest.h>
+#include <thread>
+#include <iostream>
+
+#include <fstream>
+#include <string>
+#include "src/IPC/IPC.h"
+
+// file comparison from https://stackoverflow.com/questions/6163611/compare-two-files
+#include <iterator>
+#include <string>
+#include <algorithm>
+
+#define PRINT_PIPE 0
+
+bool compare_files(const std::string&, const std::string&);
+
+void run_pipe_rx(void)
+{
+    #if PRINT_PIPE
+    std::cout << "run_pipe_rx" << std::endl;
+    #endif
+
+    char path[] = {"data/pipe_output.txt"};
+    std::fstream file;
+
+    #if PRINT_PIPE
+    std::cout << "Launch Pipe Rx" << std::endl;
+    #endif
+    CreatorIPC* pipe_rx = new CreatorPipeRx();
+    file = pipe_rx->openWriteFile(path);
+
+    ReceiverIPC* pipe_file_rx = pipe_rx->createIpcRx(&file);
+    #if PRINT_PIPE
+    std::cout << pipe_file_rx->receive();
+    #else
+    pipe_file_rx->receive();
+    #endif
+}
+
+void run_pipe_tx(void)
+{
+    #if PRINT_PIPE
+    std::cout << "run_pipe_tx" << std::endl;
+    #endif
+
+    char path[] = {"data/pipe_input.txt"};
+    std::fstream file;
+
+    #if PRINT_PIPE
+    std::cout << "Launch Pipe Tx" << std::endl;
+    #endif  
+    
+    CreatorIPC* pipe_tx = new CreatorPipeTx();
+    file = pipe_tx->openReadFile(path);
+    
+    SenderIPC* pipe_file_tx = pipe_tx->createIpcTx(&file);
+    #if PRINT_PIPE
+    std::cout << pipe_file_tx->send();
+    #else
+    pipe_file_tx->send();
+    #endif
+}
+
+int run_pipe_test(void)
+{
+    #if PRINT_PIPE
+    std::cout << "----------------------Starting run_pipe_test----------------------" << std::endl;
+    #endif
+    std::thread th_rx(run_pipe_rx);
+    std::thread th_tx(run_pipe_tx);
+
+    //wait for both threads to finish
+    th_rx.join();
+    th_tx.join();
+
+    #if PRINT_PIPE
+    std::cout << "----------------------Ending run_pipe_test----------------------" << std::endl;
+    #endif
+    return 0;
+}
+
+TEST(IpcTest, Pipe)
+{
+    // run pipe file transfer
+    run_pipe_test();
+
+    // compare input.txt and output.txt
+    std::string input_path{"data/pipe_input.txt"};
+    std::string output_path{"data/pipe_output.txt"};
```

Comment:
As this file is pushed into repo, I worry in case 'run_pipe_test' just does nothing, this test will pass

Reply:
I will clear `pipe_output.txt` to make sure it actually copies the file over correctly.

Reply:
Ok, so I think it's actually okay in the way it is currently.

When opening a file for writing I'm already clearing the contents (`IPC.cc`, line 13):

`file.open(path, std::fstream::out | std::fstream::trunc);`

[clear data inside text file in c++](https://stackoverflow.com/questions/17032970/clear-data-inside-text-file-in-c)

Reply:
That's right, but IPC.cc is the code which is tested. So, from test point of view this can still pass if copying failed. Actually I'd expect this test to pass if that file.open failed